### PR TITLE
update react-redux connect options

### PR DESCRIPTION
**Pull request by dannycochran**
The existing Options had a redundant "withRef" parameter (it was already inheriting from ConnectOptions), and were missing some helper functions for diffing state and props:

https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/react-redux/index.d.ts [hunk 1]
```diff
@@ -173,11 +173,30 @@ interface Options extends ConnectOptions {
...
+     */
+    areStatePropsEqual?: (nextStateProps: TStateProps, prevStateProps: TStateProps) => boolean;
+
     /**
-     * If true, stores a ref to the wrapped component instance and makes it available via
-     * getWrappedInstance() method. Defaults to false.
+     * When pure, compares the result of mergeProps to its previous value.
+     * @default shallowEqual
      */
-    withRef?: boolean;
```
Comment (thasner) [inline types/react-redux/index.d.ts]: why was this removed?

**Comment by thasner** [comment]
Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17584

File: types/react-redux/index.d.ts (hunk 1 above)
Comment (thasner) [inline types/react-redux/index.d.ts]: looks like it should be moved inside of [`ConnectOptions`](https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments-1)

Comment (thasner) [inline types/react-redux/index.d.ts]: From regular [`connect`](https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options)
> [options] (Object) If specified, further customizes the behavior of the connector. In addition to the options passable to connectAdvanced() (see those below), connect() accepts these additional options

Comment (dannycochran) [inline types/react-redux/index.d.ts]: It's already inside of ConnectOptions.

Comment (thasner) [inline types/react-redux/index.d.ts]: ah, excellent, sounds good

**Review by thasner — APPROVED** [review]
just gotta fix the build

File: types/react-redux/index.d.ts [hunk 2]
```diff
@@ -164,7 +164,7 @@ interface MergeProps<TStateProps, TDispatchProps, TOwnProps, TMergedProps> {
     (stateProps: TStateProps, dispatchProps: TDispatchProps, ownProps: TOwnProps): TMergedProps;
 }
 
-interface Options extends ConnectOptions {
+interface Options<TStateProps = any, TOwnProps = any, TMergedProps = any> extends ConnectOptions {
```
Comment (clayne11) [inline types/react-redux/index.d.ts]: Why default to `any`? I would argue that this is dangerous and can give people a feeling of safety when in actuality `any` has been inferred without their knowledge.

My preference is not to default this at all, and if it cannot be inferred, it will be inferred to `{}` which will force people to deal with the issue.

If you want `any` you should have to declare that explicitly.

**Review by clayne11 — CHANGES_REQUESTED** [review]

File: types/react-redux/index.d.ts (hunk 2 above)
Comment (thasner) [inline types/react-redux/index.d.ts]: why give `TStateProps` and `TOwnProps` defaults?

Comment (thasner) [inline types/react-redux/index.d.ts]: ^^ ignore that, default to `{}` makes sense

Comment (dannycochran) [inline types/react-redux/index.d.ts]: @clayne11 @thasner the build was failing because the arguments were required, and next-redux-wrapper imports Options:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/next-redux-wrapper/index.d.ts#L21

I'd prefer we default these for now, fix next-redux-wrapper in a separate PR, then remove the "any" typings.

Comment (clayne11) [inline types/react-redux/index.d.ts]: I'm pretty sure that generic arguments are never required in TS - they get inferred by the compiler if they're not provided.

If I'm wrong, I'd prefer the default is to `{}` not to `any` since that's a much safer, more conservative solution.

Comment (dannycochran) [inline types/react-redux/index.d.ts]: I think it may depend on your compiler options -- but the build was failing (I don't think I can link to it?) due to the arguments not being provided.

I'll change the default to {}

**Review by clayne11 — APPROVED** [review]

**Review by thasner — APPROVED** [review]

**Comment by Razinsky** [comment]
There's an issue with the current Options typings. It never passes the State and `areStatesEqual` needs it!

**Comment by ThomasMichon** [comment]
Yes, references to `Options` should be passing as `Options<State, TStateProps, TOwnProps, TMergedProps>`. Right now they're ommitting `State`, causing `State` to get treated like `TStateProps` and thus failing when you try to use `areStatesEqual` or similar.
